### ISSUE TITLE
[Question]: Same item found in different lists - how can I idiomatically solve the "shared state" problem?

### ISSUE THREAD

**Issue by lucavenir** [rrousselGit/riverpod#3285]
TL;DR: in a Flutter application, we have different paginated lists which can possibly _overlap_, i.e. an `Item` could appear in `myItems(page: 2)`, `favoriteItems(page: 0)`, `searchedItems(page: 7)`, ...
We want a idiomatic, testable and clean way to synchronize shared state (e.g. `isFavorite` state) for all of the above *without adding extra BE requests*.

There's a [complete example](https://github.com/lucavenir/riverpod_paginated_views) available to reproduce the question above.

If you're reading this, you can ignore the following wall of text. Originally this issue proposed an imperative API, which is rejected by Riverpod's declarative nature.

--- 

~# Is your feature request related to a problem? Please describe.~
~I find myself, very often within my daily work, in need of a way to:~

~1. synchronize state in the whole application, **whose initialization should be made from another provider**~
~2. initialize a provider with an **already-computed** value (e.g. from another side effect, hard coded from the UI, from ephemeral state, from another provider, etc.) ~

~These two are related, but not also not quite.~
~### 1. synchronization~
~About (1), I've written a [complete example](https://github.com/lucavenir/riverpod_paginated_views) which shows what we've been discussing so far. Running that app shows how an extra request per item is required.
Anyway, here's the requirements summary:~

~> There are some items, which are remotely fetched. It is possible to favorite/unfavorite them~
~> The items can be displayed in a paginated and possibly infinite list view, in a page. Actually, many of those.~
~> The items can be obtained in different ways: from a home page, from a search page or from a favorite page~
~> It is possible to fetch the details of a single item in their related details page~
~> In all of the above cases, we want to obtain the `int? favoriteId` state, which tells whether the item is favorited or not~
~> This important piece of information **must stay synchronized** across all pages, and across all the items, details included~

~Turns out, this is not so easy as of now.~
~In the repo above, I've tried an approach we've discussed on discord.~
~The main takeaway is: there is no easy way to pass in state from a provider to its "child" (sort-to-say).~
~I would call this problem "eager initialization with value", because I'd love to override a provider with a value that I already have, from its "parent".~

~### 2. initializeWith~
~About (2), I did not write a full example (yet), but I can describe the problem quickly.~

~> In the aformentioned application, I want to be able to search for items via some filters~
~> Following riverpod's best practices, filters are controlled via a notifier, and search results depend on them~
~> Results are pushed imperatively (see the structure below)~
~> **Filters can be saved for later use** (aka sent remotely to a DB) - it's handy to have them stored somewhere (e.g. user-generated content)~
~> Filters can also be obtained from other parts of the application (e.g. notifications, deep links, etc.)~
~> Summarizing, the pages can be structured as: `Home -> SearchFilters -> SearchResults`, `Home -> SavedFilters -> SearchResults`, `Home -> Notifications -> SearchResults`, etc.~
~> In other words, filters can be obtained from other parts of the application, not just *manually* by the user~
~> Therefore, I want to be able to initialize the filters notifier with the obtained filters **without triggering *double* search requests**~

~This isn't as easy as it looks.~
~The trivial observation would be to initalize the filers *before* triggering the search request, but this is not always possible.~
~In my use case, it's plain impossible. I cannot keep alive the filter controller in, say, home page, since I *need* to dispose of the filters (aka reset them) once I exit the page (or I'll get previous unintended filters when I re-enter the page).~

~# Describe the solution you'd like~
~I'd like a way to eagerly initialize, aka override, from a provider or from a widget, without using scoping or being concerned about `ProviderContainer`, at all.~
~Relying on scoping and/or the provider container feels like an internal detail leaking out of the library, rather than a way to solve a practical problem and I want to avoid that.~
~Also, scoping is not always feasible, and indeed it is not in my use case.~

~Solving (1) would look as easy as:~
```dart
@riverpod
class IsFavoriteController extends _$IsFavoriteController {
  @override
  int build(int id) => throw UnimplementedError('This provider is meant to be overridden');
}

@riverpod
FutureOr<List<Item>> someList(SomeListRef ref) async {
  final result = fetch(fromSomewhere);
  for (final item in result.items) {
    ref.override(isFavoriteControllerProvider(id), w: (ref) => item.isFavorite);
  }
}
```

~Solving (2) would look as easy as:~
```dart
class HomePage extends ConsumerWidget {
  const HomePage({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    // we want to keep these alive
    ref.listen(filtersControllerProvider, (_, __) {});
    return TextButton(
      onPressed: () {
        // similar could be done in the `SavedFilters` or `Notifications` pages
        ref.override(filtersControllerProvider, w: (ref) => Filters.empty());
        // SearchResultsRoute().go(context);
      },
      child: const Text('Logout'),
    );
  }
}
```

~But obviously imagining is way harder than actually implementing it, so I'm not sure if this is even possible.~

~# Describe alternatives you've considered~
~Yes, I know, "one single data flow". But these words are not helpful in the above examples. I cannot distill the above requirements within the "one simple data direction" mantra.~

~Don't get me wrong, to me, Riverpod is the perfect tool for most of my use cases; but this one is a bit tricky.~
~I'm not sure yet, if this is a problem with the way riverpod works or if I'm missing a piece in how to use Riverpod for the above usecases.~
~My hope is - obviously - the latter.~

~So the alternatives are, for point (1), performing an extra network request to fetch the favorite state of each item, which is not ideal.~
~In my use case it rapidly begins to be not feasible.~

~For point (2), the obvious non-enthusiastic solution is... associate providers / controller with use cases. You'd write a filter controller for each use case (from home, from saved, from notification, etc.); search results should also be duplicated. We could minimize duplication via mixins or common functions, but it gets *hard* to maintain.~

~Another workaround I've thought to - a smelly one! - is the following~
```dart
extension RefKeepAliveWhileExecuting<T> on AutoDisposeFutureProviderRef<T> {
  Future<T> keepAliveWhileExecuting(AsyncValueGetter<T> callback) {
    final link = keepAlive();
    try {
      return callback();  // e.g asynchronously push other pages after we initialize with something, etc.
    } finally {
      link.close();
    }
  }
}
```
~AKA we keep alive the filter controller until the `push` is done, but this is heavily bug prone.~
~For example, when using `go_router`, there is no guarantee that `push.future` will ever finish (e.g. using `context.go`).~

~**Additional context**~
~Is it possible to see expect any change about the above in the near future (aka 3.0)? :eyes:~

**Comment by rrousselGit** [comment]
> The main takeaway is: there is no easy way to pass in state from a provider to its "child" (sort-to-say).

What does this mean exactly?
Isn't that what providers do, passing state to their children?

**Comment by rrousselGit** [comment]
> Solving (1) would look as easy as:
> 
> ```dart
> @riverpod
> class IsFavoriteController extends _$IsFavoriteController {
>   @override
>   int build(int id) => throw UnimplementedError('This provider is meant to be overridden');
> }
> 
> @riverpod
> FutureOr<List<Item>> someList(SomeListRef ref) async {
>   final result = fetch(fromSomewhere);
>   for (final item in result.items) {
>     ref.override(isFavoriteControllerProvider(id), w: (ref) => item.isFavorite);
>   }
> }
> ```

Why can't you do:

```dart
class IsFavoriteController extends _$IsFavoriteController {
  @override
   bool build(int id) => ref.watch(someListProvider).items.firstWhereOrNull((i) => i.id == id)?.isFavorite ?? false;
}
```

**Comment by lucavenir** [comment]
> Honestly there's a bit too much to unpack in that issue.

Okay - true - give me one more chance - let's focus just on part (1) for now if you will

> Why can't you do [...]

I forgot to mention that `someListProvider` is paginated - every list in these example is, even the favorite one.
1. what's the correct page to watch? Should I pass `page` around everywhere - which is not really feasible for my use case. (`page` leaks everywhere - it not just tedious but... you quickly find yourself with widgets or provider needing a `page` they don't really care about)
2. accessing lists is generally unsafe - especially the `?? false` part: you can never really rely on `firstWhere`, e.g. invalidation refresh the provider and removes an item, or any other unexpected change. Generally speaking, the mental reactive model I have should watch this detail and not the other way around

> It's not immediately clear to me what we're trying to solve.

I'm trying my best to summarize my use case, but without the right constraints, riverpod *has* a solution for the above problem. *With* the right requirements, I find myself in a pinch and I'm not sure this issue is open in here.
Give me one more chance with a bullet point:
- You want to handle several list of items, paginated, possibly infinite, that have some common properties that *need to be in sync!*
- I can't find an idiomatic way to "always find myself (aka app state) in sync between multiple providers (or controllers)"
- "in sync" includes and isn't limited to the following events: some items are removed from the favorites (they disappear from a list and their favorite status should be propagated everywhere), some items are removed from the favorites (they appear in the favorite list, propagate it), some items are deleted entirely (etc.)

> Your example has way too much code

Yes. Sorry.
Don't look at the code - execute the trivial app and notice how an extra request is performed when you tap a like, then - if you may - read the summary above to understand why I must perform that to synchronize state everywhere.

> But no matter what, imperative overrides are a no-go.
Honestly I'm fine with whatever. Procedural, Functional, Object-Oriented, Event-based. Fine.
I need to ship, ship, ship features. Test them. And "fast" (aka at the right speed).
I'm all-in for reactivity - I guess you know 😛 - but I can't solve this problem.
Maybe a reactive `emit`

**Comment by rrousselGit** [comment]
> 1. what's the correct page to watch? Should I pass `page` around everywhere

Sure, pass it around if need be.
If may be inconvenient by your standards, but it works.  
I know that you dislike props drilling, as we've discussed variants of this multiple times. But that's ultimately the current solution.


___

I'm open to proposals. But I don't think this specific feature request is realistic for technical reasons.
If you have other ideas, I'm open to investigate those.

Although you always have the possibility to make a Notifier with a public setter in it, and invoke that setter wherever you want.
You'll break a few Riverpod principles and may have to use workarounds to bypass the internal error handling, but that should work.

**Comment by lucavenir** [comment]
Hi, I'm back. I wanted to reason about and experiment on this before wasting your time.

> Sure, pass it around if need be.

I don't mind passing around props anymore, since our last discussion I accepted that 😄 I tried doing that, but it doesn't solve the problem.

You proposed:
```dart
class IsFavoriteController extends _$IsFavoriteController {
  @override
   bool build(int id) => ref.watch(someListProvider).items.firstWhereOrNull((i) => i.id == id)?.isFavorite ?? false;
}
```

But this won't work, because:

1.  item could be in `searchProvider(page: 5)`, into `myFavoritesProvider(page: 0), and into `myListProvider(page: 1)`. See? Passing `page` around gives inconsistent states, because `.firstWhereOrNull` will return `null` in most cases, thus `false` even if it's favorited
2. Say we solve (1), somehow. Again, we have `myListProvider`, `myFavoritesProvider`, `searchProvider`, `homeProvider`, `anotherListProvider`. They're all asynchronous `List<T>` fetchers. Question: which one should `IsFavoriteController` watch?

Then you propose to use a notifier just to have a public setter in it, but as you've pointed out that clearly breaks riverpod usage.

**Comment by lucavenir** [comment]
I also changed the first comment so this issue is narrowed down to a clear and simple problem. I hope this helps.

**Comment by mattermoran** [comment]
This applies to essentially any app out there. And it's very common to see inconsistent state in the app because it's changed in one place but not other (especially inside lists). 

The way I handle it is by essentially having a single provider per `Item`. And instead of rendering `List<Item>` you render `List<Reference<Item>> so the list is always up to date.  In this case my whole app is synchronized. Doesn't matter if I have 100 different lists where that item is present. They all will render exactly same thing. And once one of the properties is updated like `isFavorite` because button was clicked or just because new state comes in a new list that will be reflected everywhere. 

As to how exactly to do this is up to you. 

In my case I use graphql, hook into response coming from api, and then place it into my cache that is then read by a provider. 

This is a very basic explanation as I do bunch of other stuff like normalizing data to handle deeply nested objects, merging data so incompletely data doesn't just override, etc. 

The comes with bunch of other advantages like navigating around the app with just ids as you always know where to get that data from, "free" offline mode as you already have the data cached, etc

But you can keep is simple and skip normalization if you have simple enough app and use-case. 

I'm sure I can make it into a package. It's very generic and technically could be used with any "provider" package out there be it riverpod, bloc, or anything

**Comment by rrousselGit** [comment]
Note that 3.0 is already adding a way to listen a provider without mounting it:
```dart
ref.listen(
  provider,
  weak: true,
  (prev, next) {

});
```

This should help solving this sort of problem. An `Item` provider could listen to multiple providers emitting an `Item ` instance and pick the most recent one, without triggering those providers.

**Comment by mattermoran** [comment]
Having generated repository for entities sounds fantastic and most likely cover all those cases

**Comment by lucavenir** [comment]
Sorry bout the late relpy! For @mattermoran:
> As to how exactly to do this is up to you.

Well ofc. My point is: It seems I can't find an idiomatic way to do so.

> In my case I use graphql, hook into response coming from api, and then place it into my cache that is then read by a provider.

See? That's the point. Riverpod is [A Reactive Caching and Data-binding Framework](https://riverpod.dev/). So why must we re-implement our own cache, again? (I suppose [via a database / persistent storage of some kind](https://isar.dev/)).

Anyways, your following comment brings up a suggestion for @rrousselGit:

> This is a very basic explanation as I do bunch of other stuff like normalizing data to handle deeply nested objects, merging data so incompletely data doesn't just override, etc.
> This comes with bunch of other advantages like navigating around the app with just ids as you always know where to get that data from, "free" offline mode as you already have the data cached, etc

This should be addressed in the documentation: if Riverpod is meant to be used with some data normalization, or any best practice really, it should be explicitly told in a dedicated documentation page.

**Comment by lucavenir** [comment]
And about riverpod 3's new API @rrousselGit just showed: again, I'm not sure how this solves the above problem.

The main pain point of this issue is: even after data normalization, I am unable to determine which provider should be watched. Here you're suggesting something like:

```dart
watched.valueOrNull?.firstWhereOrNull((item) => item.id == id);  // weak listen, but still referencing one single list
```

But as I wrote 2 weeks ago (my last message), because of points (1) and (2), this proposal is cool, but I can't see how it addressed the "synced state problem".

Because of data normalization, I would love to do the following instead:

```dart
@Riverpod(lazy: true)
int favoriteId(int id) {
  return Uninitialized<int>;  // sentinel value, or something similar
}

// somewhere else
@riverpod
List<int> someList() async {
  final elements = await fetch();

**Comment by rrousselGit** [comment]
With normalized data, you have only a single source of truth for what an entity is.

There's no case of "`Item` in home vs `Item` in detail page". They all should be using the exact same instance.
There shouldn't be a case where mutating one `Item(id: 42)` isn't reflected on other `Item(id: 42)`, since there's only a single instance of it.

**Comment by lucavenir** [comment]
> There shouldn't be a case where mutating one Item(id: 42) isn't reflected on other Item(id: 42), since there's only a single instance of it.

Yes this is very clear. This is what we're aiming at. This is why we're defining this `ItemController` with the "change favorite status mutation" on it.
The question in this thread is: **how?** (i.e. with riverpod)
The `.firstWhereOrNull` solution might bring to inconsistent state - if I've understood this correctly - so it's not really viable. Also, which list and which page should I `watch`?.

> They all should be using the exact same instance
> There's no case of "Item in home vs Item in detail page"

I think I might have a know-how gap I might need to fill.
Please, I'd be thankful if you'd help me out with this / point me towards some readings afterwards.
_But._

**Here's my POV, with a pragmatic problem to solve:**

These two lists do come from semantically different sources, tho. Say I set to favorite _my own item_.
Such item now belongs to `faovritesProvider(page: 5)`, while being on `myItemsProvider(page: 2)`, but also casually appearing on `homeProvider(page: 0)`.

The only way - that I currently see - is to have "one single source of truth" is to have an `ItemController`, but - again - the problem is its initialization. Am I stuck in a loop?

**Comment by lucavenir** [comment]
Hi 😸

I'm back on this issue, I re-read the whole conversations. Me and some colleagues tried [some solutions](https://github.com/lucavenir/riverpod_paginated_views/pulls?q=) towards this "synced-state-problem", but nothing is really working out.

I'm just curious - is this really a "riverpod" problem? Or is it a more generalized problem?
Because I can see our issue reproduced on several social platforms, e.g. Twitter's like system.

So I'm wondering if riverpod is ever considering to find a solution for this problem.
Or if it's relevant to riverpod at all.
Answering this would greatly help our R&D process. We might change everything just to tackle this one out.
And, consequently, we might close this issue with a "wont-fix" label.

**Comment by rrousselGit** [comment]
The problem is by no means unique to Riverpod. It's a general software issue

That doesn't mean Riverpod won't try to solve this. That's what Riverpod does anyway, trying to solve general software issues. That's why Riverpod also will add things like offline persistence & stuff. Not unique to Riverpod, but good to have

**Comment by lucavenir** [comment]
Thank you for replying so _fast_!

I'm glad to know the issue is acknowledged then 😃

**Comment by rrousselGit** [comment]
Using the upcoming `listen(weak: true)`, you would become able to do:

```dart
@riverpod
Item itemById(ItemByIdRef ref) {
  ref.listen(
    homePageItems(page: 0),
    weak: true,
    (_, next) {
    
    }
}

**Comment by lucavenir** [comment]
Hello there! First of all thank you for solving this.
While we wait for Riverpod 3, I still can't fully understand how to solve this, given I can use the new `weak: true` API.

It's been a while, so here's the TLDR of the problem, hoping that you could clarify this.

_An app shows a list of `Book`s in its home page, but also allows me to search for a `Book` in a search page. Then again, each book has its detail page. Finally, there's a Library page, showing different libraries, each with their list view of `Book`s._

Here's the twist: **every book is likeable**, aka all entities here share this common property.
How can I properly keep track of a book's `bool isFavorite` property, in every different page?
A typical user flow is, e.g., a book is liked in the library page; pressing "back", the home shows this book's favorite state updated accordingly.

How can I solve this problem using `listen: weak`?

**Comment by w0rsti** [comment]
As we are basically talking about related issues (#3781) I am currently facing the same issue. 

I think how a weak listener can help is as follows (sorry I am on the phone so no code examples):

All mutations on an item should happen in the detailed provider/notifier. E.g. your update function on to like an item should be within the ItemNotifier. 

This way all other providers can weakly listen to the according provider and update their internal state to match it to the ItemNotifier one. 

So your PaginatedProvider fetched a List<Item> which then listens weakly to every ItemNotifier and if this notifier exposes a new state of an item in its list, it will update its state (the list of item) with the updated item.
